### DISCUSSION TITLE
Nuxt SSR Cannot read property 'toLowerCase' of undefined

### DISCUSSION THREAD

**Discussion by brignolij** [nuxt/nuxt#9557] [Questions (Nuxt 2)]
### Versions

- nuxt: 2.15.7
- @nuxtjs/composition-api: 0.24.7
- node: 14.17.1

### Reproduction

Hi, there is no reproduction. the issue appear on a "large project", i give the most information possible.

<details open>
<summary>Additional Details</summary>
<br>

</details>

### Steps to reproduce
When deploying Nuxt SSR using composition-api and typescript on AWS on Docker instance, I  get some "Server error" when trying to get index page.

[buildspec.yml.txt](https://github.com/nuxt/nuxt.js/files/6814467/buildspec.yml.txt)
[docker-compose.production.yml.txt](https://github.com/nuxt/nuxt.js/files/6814468/docker-compose.production.yml.txt)
[Dockerfile.txt](https://github.com/nuxt/nuxt.js/files/6814469/Dockerfile.txt)
[package.json.txt](https://github.com/nuxt/nuxt.js/files/6814471/package.json.txt)
[tsconfig.json.txt](https://github.com/nuxt/nuxt.js/files/6814473/tsconfig.json.txt)
[nuxt.config.ts.txt](https://github.com/nuxt/nuxt.js/files/6814514/nuxt.config.ts.txt)

### What is Expected?
No error

### What is actually happening?
Nuxt server is having an issue. i get the following error on each home page rendering :

>  ERROR  Cannot read property 'toLowerCase' of undefined
> 
>   at getLocation (codebuild/output/src099372309/src/node_modules/vue-router/dist/vue-router.common.js:2638:0)
>   at new HTML5History (codebuild/output/src099372309/src/node_modules/vue-router/dist/vue-router.common.js:2550:0)
>   at new VueRouter (codebuild/output/src099372309/src/node_modules/vue-router/dist/vue-router.common.js:2914:0)
>   at createRouter (.nuxt/router.js:485:0)
>   at createApp (.nuxt/index.js:70:0)
>   at module.exports.__webpack_exports__.default (.nuxt/server.js:81:0)
>   at /app/node_modules/vue-server-renderer/build.prod.js:1:78890
>   at new Promise (<anonymous>)
>   at /app/node_modules/vue-server-renderer/build.prod.js:1:78362
>   at Object.renderToString (/app/node_modules/vue-server-renderer/build.prod.js:1:81897)

I already tried to clean plugin, middleware to detect the error, but i didn't found it.

### About deployment
When developing using nuxt serve and when testing production on local and on local docker, we cannot reproduce the issue.
We deploy by this way:
On Ubuntu (AWS Codebuild) server we make "yarn install" and "nuxt build"
then we take the result in zip and push it to AWS Elastic beanstalk Docker server. And we use docker based on node:14.17.1 to run nuxt start.

Can you help me to understand, from the server error, which part of nuxt this error  comes from or/and how to debug it ?
[buildspec.yml.txt](https://github.com/nuxt/nuxt.js/files/6814492/buildspec.yml.txt)
[docker-compose.production.yml.txt](https://github.com/nuxt/nuxt.js/files/6814493/docker-compose.production.yml.txt)
[Dockerfile.txt](https://github.com/nuxt/nuxt.js/files/6814494/Dockerfile.txt)
[package.json.txt](https://github.com/nuxt/nuxt.js/files/6814495/package.json.txt)
[tsconfig.json.txt](https://github.com/nuxt/nuxt.js/files/6814496/tsconfig.json.txt)
[nuxt.config.ts.txt](https://github.com/nuxt/nuxt.js/files/6814514/nuxt.config.ts.txt)

**Comment by asadullah009** [comment]
The stack trace shows the crash happens during SSR router creation inside `vue-router`:

```txt
at getLocation (vue-router.common.js:2638:0)
at new HTML5History
at new VueRouter

getLocation() internally expects a valid location/base value and calls .toLowerCase() on it. One of those values is becoming undefined only in your AWS production environment.

Since:

local production works
local Docker works
only AWS SSR fails

this strongly suggests an environment/runtime issue rather than a Nuxt page/component bug.

The first thing I would check is any runtime config using environment variables, especially router config:

router: {
  base: process.env.ROUTER_BASE
}

or:

publicRuntimeConfig: {
  apiUrl: process.env.API_URL
}

If one of these values is undefined in AWS, vue-router can fail during SSR initialization.

Another important point is your deployment flow:

AWS CodeBuild:
- yarn install
- nuxt build

then copy artifacts into another Docker image
then run nuxt start

This is not ideal for SSR applications. Build and runtime environments should usually be identical.

A safer approach is building directly inside the same Docker image that runs the app:

FROM node:14.17.1

WORKDIR /app

COPY . .

RUN yarn install
RUN yarn build

CMD ["yarn", "start"]

Also check whether AWS/Elastic Beanstalk/reverse proxy is forwarding request headers correctly.

Add temporary logging:

export default function ({ req }) {
  if (process.server) {
    console.log(req.url)
    console.log(req.headers)
  }
}

and verify that headers like these always exist:

host
x-forwarded-host
x-forwarded-proto

because vue-router SSR depends on request location information.

I do not think @nuxtjs/composition-api is the direct cause here, because the crash occurs before page rendering, during router creation itself.

So the most likely causes are:

Undefined environment variable used in router/runtime config
AWS proxy/header issue
Build/runtime mismatch between CodeBuild and Docker runtime
Middleware modifying req.url or headers
